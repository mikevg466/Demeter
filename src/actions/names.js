import types from './types';
import { shuffle } from '../utils/array';
import NAMES from '../assets/data/names.json';
// import NAMES from '../assets/data/namesCustom.json';

export const addCustomName = () => ({
  type: types.ADD_CUSTOM_NAME,
});

/**
 * Action creator for DELETE_PREVIEW_NAME reducer
 * @param {Number} index The index of the array item to delete
 * @return {Object} action with names array
 */
export const deletePreviewName = index => ({
  type: types.DELETE_PREVIEW_NAME,
  index,
});

/**
 * Action creator for INITIALIZE_SORTING_LIST reducer
 * @return {Object} action with names array
 */
export const _initializeSortingList = () => ({
  type: types.INITIALIZE_SORTING_LIST,
});

/**
 * Action creator for RESET_NAMES reducer
 * @return {Object} action with names array
 */
export const resetNames = () => ({
  type: types.RESET_NAMES,
});

export const setCustomName = name => ({
  type: types.SET_CUSTOM_NAME,
  name,
});

/**
 * Action creator for SET_GENDER_TYPE reducer
 * @param {String} genderType Gender type constant: M for male, F for female
 * @return {Object} action with names array
 */
export const setGenderType = genderType => ({
  type: types.SET_GENDER_TYPE,
  genderType,
});

/**
 * Action creator for SET_NAMES reducer
 * @param {Array} rawList Array of names to be set in the store
 * @return {Object} action with names array
 */
export const setNames = rawList => ({
  type: types.SET_NAMES,
  rawList,
});

/**
 * Action creator for SET_SELECTOR_NAMES reducer
 * @return {Object} action with names array
 */
export const setSelectorNames = () => ({
  type: types.SET_SELECTOR_NAMES,
});

/**
 * Action creator for SORT_SELECTOR_NAMES reducer
 * @param {Number} compareValue A positive or negative number to show
 *                                how the user wants to sort the names.
 * @param {Boolean} [removeName=false] True if the compareValue represents a
 *                                      name to remove instead of sorted
 * @return {Object} action with names array
 */
export const sortSelectorNames = (compareValue, removeName = false) => ({
  type: types.SORT_SELECTOR_NAMES,
  compareValue,
  removeName,
});

/**
 * shuffles the rawList array then splits it to create the sorting list
 * @return {undfined}
 */
export const initializeSortingList = () => dispatch => {
  dispatch(shuffleNames());
  dispatch(_initializeSortingList());
};

/**
 * Loads names from source which are shuffled and saved to the store
 * @param {String} [genderType='M'] M for male, F for female.  Defaults to M
 * @return {undfined}
 */
export const loadNames =
  (genderType = 'M') =>
    dispatch => {
      dispatch(resetNames());
      const names = NAMES.data
        .filter(({ gender }) => gender === genderType)
        .slice(0, 1000)
        .map(({ name }) => name)
        .sort();

      dispatch(setNames(names));
    };

/**
 * Removes a name from two choices and updates the sort list then sets new names
 *  to compare.
 * @param {Number} compareValue A positive or negative number to show
 *                                how the user wants to sort the names.
 * @return {undfined}
 */
export const removeName = compareValue => dispatch => {
  dispatch(sortSelectorNames(compareValue, true));
  dispatch(setSelectorNames());
};

/**
 * Selects a name from two choices and updates the sort list then sets new names
 *  to compare.
 * @param {Number} compareValue A positive or negative number to show
 *                                how the user wants to sort the names.
 * @return {undfined}
 */
export const selectName = compareValue => dispatch => {
  dispatch(sortSelectorNames(compareValue));
  dispatch(setSelectorNames());
};

/**
 * Shuffles the rawList of names then updates the state with the new list.
 * @return {undfined}
 */
export const shuffleNames = () => (dispatch, getState) => {
  const {
    names: { rawList },
  } = getState();
  dispatch(setNames(shuffle(rawList)));
};
