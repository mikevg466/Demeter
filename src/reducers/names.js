import update from 'immutability-helper';
import types from '../actions/types';
import { splitArray } from '../utils/array';

export const namesInitialState = {
  rawList: [],
  sortingList: [],
  currentSortList: [],
  finalList: [],
  firstName: '',
  secondName: '',
  mainIdx: 0,
  leftIdx: 0,
  rightIdx: 0,
  isSortingFinished: false,
  genderType: '',
};

/**
 * The main names reducer function to be exported
 * @param  {Object} [state=namesInitialState]
 *    The current names store to be manipulated.
 *    If null, then the default initial state.
 * @param  {Object} action
 *    The reducer action, containing the action type and data to be processed.
 * @return {Object} The updated names state.
 */
export default function namesReducer(state = namesInitialState, action) {
  switch (action.type) {
    case types.INITIALIZE_SORTING_LIST: {
      const sortingList = splitArray(state.rawList);
      return update(state, {
        sortingList: { $set: sortingList },
      });
    }
    case types.RESET_NAMES:
      return update(state, {
        rawList: { $set: namesInitialState.rawList },
        sortingList: { $set: namesInitialState.sortingList },
        currentSortList: { $set: namesInitialState.currentSortList },
        finalList: { $set: namesInitialState.finalList },
        firstName: { $set: namesInitialState.firstName },
        secondName: { $set: namesInitialState.secondName },
        mainIdx: { $set: namesInitialState.mainIdx },
        leftIdx: { $set: namesInitialState.leftIdx },
        rightIdx: { $set: namesInitialState.rightIdx },
        isSortingFinished: { $set: namesInitialState.isSortingFinished },
      });
    case types.SET_GENDER_TYPE:
      return update(state, {
        genderType: { $set: action.genderType || '' },
      });
    case types.SET_NAMES:
      return update(state, {
        rawList: { $set: action.rawList || [] },
      });
    case types.SET_SELECTOR_NAMES: {
      const { sortingList, mainIdx, leftIdx, rightIdx, finalList } = state;
      if (!sortingList.length) return state;

      let nextFinalList = finalList;
      const isSortingFinished =
        !sortingList[mainIdx] || !sortingList[mainIdx + 1];

      let firstName;
      let secondName;
      if (!isSortingFinished) {
        firstName = sortingList[mainIdx][leftIdx];
        secondName = sortingList[mainIdx + 1][rightIdx];
      }
      if (isSortingFinished) {
        nextFinalList = sortingList[0].slice();
      }
      return update(state, {
        isSortingFinished: { $set: isSortingFinished },
        firstName: { $set: firstName || '' },
        finalList: { $set: nextFinalList },
        secondName: { $set: secondName || '' },
      });
    }
    case types.SORT_SELECTOR_NAMES: {
      const { compareValue, removeName } = action;
      let { leftIdx, rightIdx, mainIdx, sortingList, currentSortList } = state;
      currentSortList = currentSortList.slice();

      if (compareValue < 0) {
        if (!removeName) {
          currentSortList.push(sortingList[mainIdx][leftIdx]);
        }
        leftIdx++;
      } else if (compareValue > 0) {
        if (!removeName) {
          currentSortList.push(sortingList[mainIdx + 1][rightIdx]);
        }
        rightIdx++;
      }

      const isCurrentSortFinished =
        leftIdx >= sortingList[mainIdx].length ||
        rightIdx >= sortingList[mainIdx + 1].length;

      if (isCurrentSortFinished) {
        sortingList = sortingList.slice();

        if (leftIdx < sortingList[mainIdx].length) {
          currentSortList = [
            ...currentSortList,
            ...sortingList[mainIdx].slice(leftIdx),
          ];
        }
        if (rightIdx < sortingList[mainIdx + 1].length) {
          currentSortList = [
            ...currentSortList,
            ...sortingList[mainIdx + 1].slice(rightIdx),
          ];
        }

        sortingList.splice(mainIdx, 2, currentSortList.slice());
        currentSortList = [];
        mainIdx++;
        leftIdx = 0;
        rightIdx = 0;
        if (sortingList.length - mainIdx < 2) {
          mainIdx = 0;
        }
      }

      return update(state, {
        leftIdx: { $set: leftIdx },
        rightIdx: { $set: rightIdx },
        mainIdx: { $set: mainIdx },
        sortingList: { $set: sortingList },
        currentSortList: { $set: currentSortList },
      });
    }
    default:
      return state;
  }
}
