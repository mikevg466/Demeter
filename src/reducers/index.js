import { combineReducers } from 'redux';
import names from './names';
import user from './user';

// Use ES6 object literal shorthand syntax to define the object shape
const rootReducer = combineReducers({
  names,
  user,
});

export default rootReducer;
