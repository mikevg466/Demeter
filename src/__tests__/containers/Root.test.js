import React from 'react';
import { shallow } from 'enzyme';
import {
  RootContainer,
  mapStateToProps,
  mapDispatchToProps,
} from '../../containers/Root';
import { namesInitialState } from '../../reducers/names';

const mockState = {
  genderType: namesInitialState.genderType,
  isSortingFinished: namesInitialState.isSortingFinished,
  rawList: namesInitialState.rawList,
  sortingList: namesInitialState.sortingList,
};
const mockDispatchers = {
  loadNames: jest.fn(),
  setGenderType: jest.fn(),
  setSelectorNames: jest.fn(),
};
const mockProps = {
  ...mockState,
  ...mockDispatchers,
};

describe('Root Container', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<RootContainer {...mockProps} />);
    expect(wrapper.debug()).toMatchSnapshot();
  });

  describe('mapStateToProps', () => {
    it('outputs the expected state when passed a mock store', () => {
      const mockStore = {
        names: namesInitialState,
      };

      expect(mapStateToProps(mockStore)).toEqual(mockState);
    });
  });

  describe('mapDispatchToProps', () => {
    expect(Object.keys(mapDispatchToProps)).toEqual(
      // eslint-disable-next-line comma-dangle
      Object.keys(mockDispatchers)
    );
  });
});
