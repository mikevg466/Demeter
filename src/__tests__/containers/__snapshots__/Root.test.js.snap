// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root Container renders correctly 1`] = `"<RootComponent displayListPreview={true} genderType=\\"\\" isLoading={true} isSortingFinished={false} setGenderType={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} />"`;
